From 691cb432aeb7828f1b7e42005725621716daac7b Mon Sep 17 00:00:00 2001
Message-Id: <691cb432aeb7828f1b7e42005725621716daac7b.1423872525.git.feitong.yi@intel.com>
In-Reply-To: <3062c7e6b919c393846ff0e13558e7f648b352ca.1423872525.git.feitong.yi@intel.com>
References: <3062c7e6b919c393846ff0e13558e7f648b352ca.1423872525.git.feitong.yi@intel.com>
From: Shashank Sharma <shashank.sharma@intel.com>
Date: Wed, 28 Jan 2015 13:28:13 +0530
Subject: [PATCH 25/95] MUST_REBASE [VPG]: adf/intel/hdmi: HDMI port
 programming

This patch adds a new file vlv_hdmi_port.c which contains port
programming functions for HDMI display.

MUST_REBASE: Since Google ADF framework is not in upstream and also
we cannot have two display drivers, other one being i915, we need to
work on i915 and ADF convergence path before it can be upstreamed.

Issue: GMINL-5478
Change-Id: If772e187af68502c41d0d69831918df21c933bf9
Signed-off-by: Shashank Sharma <shashank.sharma@intel.com>
---
 drivers/video/adf/intel/core/vlv/Makefile          |    3 +-
 drivers/video/adf/intel/core/vlv/vlv_hdmi_port.c   |  108 ++++++++++++++++++++
 .../adf/intel/include/core/vlv/vlv_hdmi_port.h     |   43 ++++++++
 3 files changed, 153 insertions(+), 1 deletion(-)
 create mode 100644 drivers/video/adf/intel/core/vlv/vlv_hdmi_port.c
 create mode 100644 drivers/video/adf/intel/include/core/vlv/vlv_hdmi_port.h

diff --git a/drivers/video/adf/intel/core/vlv/Makefile b/drivers/video/adf/intel/core/vlv/Makefile
index 243d3ca..2f4875f 100644
--- a/drivers/video/adf/intel/core/vlv/Makefile
+++ b/drivers/video/adf/intel/core/vlv/Makefile
@@ -11,4 +11,5 @@ obj-y := vlv_dc_config.o \
 	vlv_pm.o \
 	vlv_dsi_port.o \
 	vlv_dsi_port_cmd.o \
-	vlv_pipeline.o
+	vlv_pipeline.o \
+	vlv_hdmi_port.o
diff --git a/drivers/video/adf/intel/core/vlv/vlv_hdmi_port.c b/drivers/video/adf/intel/core/vlv/vlv_hdmi_port.c
new file mode 100644
index 0000000..d1390bf
--- /dev/null
+++ b/drivers/video/adf/intel/core/vlv/vlv_hdmi_port.c
@@ -0,0 +1,108 @@
+/*
+ * Copyright © 2014 Intel Corporation
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice (including the next
+ * paragraph) shall be included in all copies or substantial portions of the
+ * Software.
+ *
+ * Author: Akashdeep Sharma <akashdeep.sharma@intel.com>
+ * Author: Shashank Sharma <shashank.sharma@intel.com>
+ */
+
+#include <drm/drmP.h>
+#include <core/vlv/vlv_dc_regs.h>
+#include <core/vlv/chv_dc_regs.h>
+#include <core/vlv/vlv_hdmi_port.h>
+#include <core/vlv/vlv_dc_config.h>
+
+u32 vlv_hdmi_port_enable(struct vlv_hdmi_port *port)
+{
+	u32 temp;
+	u32 enable_bits = SDVO_ENABLE | SDVO_AUDIO_ENABLE;
+
+	pr_info("ADF: HDMI: %s\n", __func__);
+
+	temp = REG_READ(port->control_reg);
+	temp |= enable_bits;
+
+	REG_WRITE(port->control_reg, temp);
+	REG_POSTING_READ(port->control_reg);
+
+	return 0;
+}
+
+u32 vlv_hdmi_port_prepare(struct vlv_hdmi_port *port, u32 val)
+{
+	u32 temp = REG_READ(port->control_reg);
+	temp |= val;
+
+	pr_info("ADF: HDMI: %s\n", __func__);
+	REG_WRITE(port->control_reg, temp);
+	REG_POSTING_READ(port->control_reg);
+
+	return 0;
+}
+
+u32 vlv_hdmi_port_disable(struct vlv_hdmi_port *port)
+{
+	u32 temp;
+	u32 enable_bits = SDVO_ENABLE | SDVO_AUDIO_ENABLE;
+
+	pr_info("ADF: HDMI: %s\n", __func__);
+
+	temp = REG_READ(port->control_reg);
+	temp &= ~enable_bits;
+
+	REG_WRITE(port->control_reg, temp);
+	REG_POSTING_READ(port->control_reg);
+
+	return 0;
+}
+
+bool vlv_hdmi_port_init(struct vlv_hdmi_port *port, enum port enum_port,
+		enum pipe pipe)
+{
+	pr_info("ADF: HDMI: %s\n", __func__);
+
+	switch (enum_port) {
+	case PORT_D:
+		port->control_reg = CHV_PORTD_CTRL;
+		port->adapter = intel_adf_get_gmbus_adapter(GMBUS_PORT_DPD_CHV);
+		break;
+
+	case PORT_C:
+		port->control_reg = CHV_PORTC_CTRL;
+		port->adapter = intel_adf_get_gmbus_adapter(GMBUS_PORT_DPC);
+		break;
+
+	case PORT_B:
+		port->control_reg = CHV_PORTB_CTRL;
+		port->adapter = intel_adf_get_gmbus_adapter(GMBUS_PORT_DPB);
+		break;
+
+	default:
+		pr_err("ADF: HDMI: %s: Invalid port\n", __func__);
+		return false;
+	}
+
+	port->dip_stat = VLV_AUD_CNTL_ST(pipe);
+	port->dip_ctrl = VLV_TVIDEO_DIP_CTL(pipe);
+	port->dip_data = VLV_TVIDEO_DIP_DATA(pipe);
+	port->hpd_detect = CHV_HPD_STAT;
+	port->hpd_ctrl = CHV_HPD_CTRL;
+	port->port_id = enum_port;
+	return true;
+}
+
+bool vlv_hdmi_port_destroy(struct vlv_hdmi_port *port)
+{
+	pr_info("ADF: HDMI: %s\n", __func__);
+	return true;
+}
diff --git a/drivers/video/adf/intel/include/core/vlv/vlv_hdmi_port.h b/drivers/video/adf/intel/include/core/vlv/vlv_hdmi_port.h
new file mode 100644
index 0000000..ad7cdfb
--- /dev/null
+++ b/drivers/video/adf/intel/include/core/vlv/vlv_hdmi_port.h
@@ -0,0 +1,43 @@
+/*
+ * Copyright © 2014 Intel Corporation
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice (including the next
+ * paragraph) shall be included in all copies or substantial portions of the
+ * Software.
+ *
+ * Author: Akashdeep Sharma <akashdeep.sharma@intel.com>
+ * Author: Shashank Sharma <shashank.sharma@intel.com>
+ */
+
+#ifndef _VLV_HDMI_PORT_H_
+#define _VLV_HDMI_PORT_H_
+#include <linux/types.h>
+#include <linux/i2c.h>
+#include <core/common/hdmi/gen_hdmi_pipe.h>
+
+struct vlv_hdmi_port {
+	u32 control_reg;
+	u32 dip_stat;
+	u32 dip_ctrl;
+	u32 dip_data;
+	u32 hpd_stat;
+	u32 hpd_detect;
+	u32 hpd_ctrl;
+	struct i2c_adapter *adapter;
+	enum port port_id;
+};
+
+u32 vlv_hdmi_port_enable(struct vlv_hdmi_port *port);
+u32 vlv_hdmi_port_disable(struct vlv_hdmi_port *port);
+bool vlv_hdmi_port_init(struct vlv_hdmi_port *port, enum port, enum pipe);
+bool vlv_hdmi_port_destroy(struct vlv_hdmi_port *port);
+u32 vlv_hdmi_port_prepare(struct vlv_hdmi_port *port, u32 val);
+
+#endif /* _VLV_HDMI_PORT_H_ */
-- 
1.7.9.5

